ellipticoind/src/api/helpers.rs

```diff
@@ -2,15 +2,23 @@ use crate::api::graphql::Error;
 use ed25519_zebra::VerificationKey;
 use serde::de::DeserializeOwned;
 use std::convert::TryFrom;
+use ellipticoin::PublicKey;
 
-pub fn validate_signature<D: DeserializeOwned>(cose_bytes: &[u8]) -> Result<D, Error> {
+pub fn validate_signature<D: DeserializeOwned>(cose_bytes: &[u8]) -> Result<(D, PublicKey), Error> {
     let sign1: serde_cose::Sign1 = serde_cbor::from_slice(cose_bytes)
         .map_err(|_| Error("invalid COSE message".to_string()))?;
-    let key = serde_cose::Key::from(
-        VerificationKey::try_from(<[u8; 32]>::try_from(&sign1.kid()[..]).unwrap())
+    let signer_pub_key = <PublicKey>::try_from(&sign1.kid()[..])
+        .map_err(|_| Error("invalid signature".to_string()))?;
+    let serde_key = serde_cose::Key::from(
+        VerificationKey::try_from(signer_pub_key)
             .map_err(|_| Error("invalid signature".to_string()))?,
     );
-    key.verify(&sign1)
+    serde_key.verify(&sign1)
         .map_err(|_| Error("invalid signature".to_string()))?;
-    serde_cbor::from_slice(&sign1.payload).map_err(|_| Error("invalid CBOR payload".to_string()))
+
+    let d = serde_cbor::from_slice(&sign1.payload).map_err(|_| Error("invalid CBOR payload".to_string()));
+    match d {
+        Ok(x) => Ok((x, signer_pub_key)),
+        Err(x) => Err(x)
+    }
```

Comment:
You can [map](https://doc.rust-lang.org/std/result/enum.Result.html#method.map) on the Result type so you should be able to replace this with:

```serde_cbor::from_slice(&sign1.payload)
.map((result) => (result, signer_pub_key))
.map_err(|_| Error("invalid CBOR payload".to_string()))